accumulo/src/main/java/com/yahoo/ycsb/db/accumulo/AccumuloClient.java

```diff
@@ -100,62 +104,72 @@ public void init() throws DBException {
    * table. Connect to it and if necessary, close our current connection.
    * 
    * @param t
    *          The table to open.
    */
-  public void getTable(String t) throws TableNotFoundException {
-    if (bw != null) { // Close the existing writer if necessary.
-      try {
-        bw.close();
-      } catch (MutationsRejectedException e) {
-        // Couldn't spit out the mutations we wanted.
-        // Ignore this for now.
-        System.err.println("MutationsRejectedException: " + e.getMessage());
+  public BatchWriter getWriter(String t) throws TableNotFoundException {
+    // tl;dr We're paying a cost for the ConcurrentHashMap here to deal with the DB api.
+    // We know that YCSB is really only ever going to send us data for one table, so using
+    // a concurrent data structure is overkill (especially in such a hot code path).
+    // However, the impact seems to be relatively negligible in trivial local tests and it's
+    // "more correct" WRT to the API.
+    BatchWriter writer = writers.get(t);
+    if (null == writer) {
+      BatchWriter newWriter = createBatchWriter(t);
+      BatchWriter oldWriter = writers.putIfAbsent(t, newWriter);
+      // Someone beat us to creating a BatchWriter for this table, use their BatchWriters
+      if (null != oldWriter) {
+        try {
+          // Make sure to clean up our new batchwriter!
+          newWriter.close();
+        } catch (MutationsRejectedException e) {
+          throw new RuntimeException(e);
+        }
+        writer = oldWriter;
+      } else {
+        writer = newWriter;
       }
     }
+    return writer;
+  }
 
+  /**
+   * Creates a BatchWriter with the expected configuration.
+   *
+   * @param table The table to write to
+   */
+  private BatchWriter createBatchWriter(String table) throws TableNotFoundException {
     BatchWriterConfig bwc = new BatchWriterConfig();
     bwc.setMaxLatency(
         Long.parseLong(getProperties()
             .getProperty("accumulo.batchWriterMaxLatency", "30000")),
         TimeUnit.MILLISECONDS);
     bwc.setMaxMemory(Long.parseLong(
         getProperties().getProperty("accumulo.batchWriterSize", "100000")));
-    bwc.setMaxWriteThreads(Integer.parseInt(
-        getProperties().getProperty("accumulo.batchWriterThreads", "1")));
-
-    bw = connector.createBatchWriter(t, bwc);
-
-    // Create our scanners
-    singleScanner = connector.createScanner(t, Authorizations.EMPTY);
-    scanScanner = connector.createScanner(t, Authorizations.EMPTY);
-
-    table = t; // Store the name of the table we have open.
+    final String numThreadsValue = getProperties().getProperty("accumulo.batchWriterThreads");
+    // Try to saturate the client machine.
+    int numThreads = Math.max(1, Runtime.getRuntime().availableProcessors() / 2);
+    if (null != numThreadsValue) {
+      numThreads = Integer.parseInt(numThreadsValue);
+    }
+    System.out.println("Using " + numThreads + " threads to write data");
```

Comment:
This should be on stderr of it's going to be manually printed. Otherwise it'll mess up folks using stdout to read results.